### PR TITLE
Support gamma_inc shape derivatives and fix zero-multiplier rules

### LINKED ISSUE #1310: Wrong gradients for `LogExpFunctions.xlogy` when `x == 0`
Hi!

I found that Mooncake returns wrong gradients for `LogExpFunctions.xlogy` when `x == 0`.

This issue was previously reported in https://github.com/chalk-lab/Mooncake.jl/issues/1241 but was closed again before it could be fixed.

The above issue also links to a custom rule implemented elsewhere (https://github.com/EpiAware/DistributionsInference.jl/issues/73). Maybe this can be a starting point for fixing this in Mooncake?

Below is a MWE that was written with the help of Claude:

```julia
using Mooncake, LogExpFunctions, ChainRulesCore

f(x, y) = xlogy(x, y)
rule = Mooncake.build_rrule(f, 1.0, 1.0)

grad_mooncake(x, y) = Mooncake.value_and_gradient!!(rule, f, x, y)[2][2:3]

function grad_chainrules(x, y)
    _, pullback = ChainRulesCore.rrule(xlogy, x, y)
    return pullback(1.0)[2:3]
end

grad_chainrules(0.0, 3.0) == (log(3.0), 0.0)
# true
grad_mooncake(0.0, 3.0) == (log(3.0), 0.0)
# false

grad_chainrules(0.0, Inf) == (Inf, 0.0)
# true
grad_mooncake(0.0, Inf) == (Inf, 0.0)
# false
```

### LINKED ISSUE #1312: `gamma_inc` shape derivative is `NaN` in reverse mode
**Reviewed by @seabbs before posting. Posted from a bot account, so please ping @seabbs with any questions.**

Recording the `gamma_inc` gap we have been carrying a workaround for. #1311 looks like it already fixes this, so this is here to give it a tracked issue rather than to report something new.

The shape derivative of the regularised lower incomplete gamma is `NaN` in reverse mode, at every shape we tried:

```julia
using Mooncake, SpecialFunctions
f(a) = gamma_inc(a, 2.0)[1]
rule = Mooncake.build_rrule(Tuple{typeof(f), Float64})

for a in (0.5, 1.0, 3.0, 7.0)
    g = Mooncake.value_and_gradient!!(rule, f, a)[2][2]
    fd = (f(a + 1e-6) - f(a - 1e-6)) / 2e-6
    println("a=$a  mooncake=$g  fd=$fd")
end
# a=0.5  mooncake=NaN  fd=-0.13561933520644942
# a=1.0  mooncake=NaN  fd=-0.22082542666668203
# a=3.0  mooncake=NaN  fd=-0.23184867203873694
# a=7.0  mooncake=NaN  fd=-0.00616494201651982
```

It propagates to the Gamma distribution, so any shape-parameter gradient through a Gamma CDF or log survival is `NaN`:

```julia
using Distributions
cdf(Gamma(a, 2.0), 1.5)      # NaN wrt a
logccdf(Gamma(a, 2.0), 1.5)  # NaN wrt a
```

The `x` partial is correct throughout, and `beta_inc` is correct in all three arguments.

This is not Mooncake losing a derivative. ChainRules has none to give:

```julia
ChainRulesCore.rrule(gamma_inc, 3.0, 2.0, 0)[2]((1.0, 0.0))[2]
# NotImplemented(..., derivatives of the incomplete Gamma functions with respect
#   to parameter `a` are not implemented currently:
#   https://github.com/JuliaMath/SpecialFunctions.jl/issues/317)
```

The `NaN` is Mooncake's documented masking of a `NotImplemented` partial, so the current behaviour is the honest signal rather than a bug in the strict sense. The underlying gap is upstream in JuliaMath/SpecialFunctions.jl#317, with JuliaMath/SpecialFunctions.jl#531 tracking the shape partial specifically.

#1311 adds native `gamma_inc` shape derivatives, so this should be closed by it.

Mooncake v0.5.56, SpecialFunctions v2.9.0, Julia 1.13.0.

---

For reference, our current workaround is in EpiAwareADTools.jl. The shape partial
itself is implemented in
[`src/gamma_ad.jl`](https://github.com/EpiAware/EpiAwareADTools.jl/blob/main/src/gamma_ad.jl)
as `_grad_p_a_series`, by term-by-term differentiation of the Tricomi
absolutely-convergent series, with an adaptive term budget and a measured accuracy
table. It is the same construction as Stan's `grad_reg_inc_gamma` and JAX's
`igamma_grad_a`.

It reaches Mooncake through the ChainRules rules in
[`ext/EpiAwareADToolsChainRulesCoreExt.jl`](https://github.com/EpiAware/EpiAwareADTools.jl/blob/main/ext/EpiAwareADToolsChainRulesCoreExt.jl),
lifted by `@from_chainrules` in
[`ext/EpiAwareADToolsMooncakeExt.jl`](https://github.com/EpiAware/EpiAwareADTools.jl/blob/main/ext/EpiAwareADToolsMooncakeExt.jl).

That helper is marked for deletion once an upstream rule lands, so #1311 would let
us drop it.

### PR DESCRIPTION
**Pull request by yebai**
Add native forward and reverse rules for `gamma_inc`, with stable tail derivatives. At `(a,x)=(3,50)`, the lower ratio’s shape derivative is approximately `-7.55e-19`.

Fix zero-multiplier derivatives: `∂x xlogy(0,3)=log(3)` and `∂x xexpy(0,2)=exp(2)`.

In future, consider moving Mooncake’s SpecialFunctions extension into SpecialFunctions.jl to keep differentiation rules alongside their implementations.

Closes #1310. Closes #1312. 

### Documentation Preview

Mooncake.jl documentation for PR #1311 is available at:
https://chalk-lab.github.io/Mooncake.jl/previews/PR1311/

### Performance

Performance Ratio:
Ratio of time to compute gradient and time to compute function.
Warning: results are very approximate! See [here](https://github.com/chalk-lab/Mooncake.jl/tree/main/bench#inter-framework-benchmarking) for more context.
```
┌────────────────────────────┬──────────┬──────────┬─────────────┬─────────┬─────────────┬─────────┐
│                      Label │   Primal │ Mooncake │ MooncakeFwd │  Zygote │ ReverseDiff │  Enzyme │
│                     String │   String │   String │      String │  String │      String │  String │
├────────────────────────────┼──────────┼──────────┼─────────────┼─────────┼─────────────┼─────────┤
│                   sum_1000 │ 191.0 ns │     1.52 │        1.52 │    1.37 │        7.29 │ missing │
│                  _sum_1000 │ 962.0 ns │     6.76 │        1.04 │  1060.0 │        35.1 │ missing │
│               sum_sin_1000 │  6.82 μs │     3.27 │        2.02 │    2.54 │        11.3 │ missing │
│              _sum_sin_1000 │  6.78 μs │      2.9 │        2.03 │   170.0 │        11.2 │ missing │
│                   kron_sum │ 207.0 μs │     3.82 │         3.5 │    26.2 │       345.0 │ missing │
│              kron_view_sum │ 263.0 μs │     4.13 │        3.82 │    12.5 │       330.0 │ missing │
│      naive_map_sin_cos_exp │  2.28 μs │     2.92 │         1.5 │ missing │        8.02 │ missing │
│            map_sin_cos_exp │  2.26 μs │     3.41 │        1.62 │    1.71 │        8.33 │ missing │
│      broadcast_sin_cos_exp │  2.36 μs │     3.02 │        1.51 │    4.14 │        1.73 │ missing │
│                 simple_mlp │ 374.0 μs │     4.45 │        2.52 │    1.75 │        8.68 │ missing │
│                     gp_lml │ 386.0 μs │     4.08 │         1.7 │    3.92 │     missing │ missing │
│ turing_broadcast_benchmark │  2.01 ms │     5.85 │        2.99 │ missing │        31.8 │ missing │
│         large_single_block │ 431.0 ns │     5.16 │        1.88 │  7340.0 │        30.8 │ missing │
└────────────────────────────┴──────────┴──────────┴─────────────┴─────────┴─────────────┴─────────┘
```

### CODE REVIEW

**Comment by seabbs** [comment]
Note the fixes we have been using are here: https://github.com/EpiAware/EpiAwareADTools.jl/blob/main/src/gamma_ad.jl along with associated mooncake extensions lifting into rules

The solutions here are lifted largely from stan and jax.
